src/ert/storage/local_ensemble.py

```diff
@@ -876,3 +900,127 @@ def get_response_state(
+        iens_active_index: npt.NDArray[np.int_],
+    ) -> polars.DataFrame:
+        """Fetches and aligns selected observations with their corresponding simulated responses from an ensemble."""
+        observations_by_type = self.experiment.observations
+
+        with polars.StringCache():
+            dfs_per_response_type = []
+            for (
+                response_type,
+                response_cls,
+            ) in self.experiment.response_configuration.items():
+                if response_type not in observations_by_type:
+                    continue
+
+                observations_for_type = (
+                    observations_by_type[response_type]
+                    .filter(
+                        polars.col("observation_key").is_in(list(selected_observations))
+                    )
+                    .with_columns(
+                        [
+                            polars.col("response_key")
+                            .cast(polars.Categorical)
+                            .alias("response_key")
+                        ]
+                    )
+                )
+
+                observed_cols = {
+                    k: observations_for_type[k].unique()
+                    for k in ["response_key", *response_cls.primary_key]
+                }
+
+                reals = iens_active_index.tolist()
+                reals.sort()
+                # too much memory to do it all at once, go per realization
+                first_columns: polars.DataFrame | None = None
+                realization_columns: List[polars.DataFrame] = []
+                for real in reals:
+                    responses = self._load_responses_lazy(
+                        response_type, (real,)
+                    ).with_columns(
+                        [
+                            polars.col("response_key")
+                            .cast(polars.Categorical)
+                            .alias("response_key")
+                        ]
+                    )
+
+                    # Filter out responses without observations
+                    for col, observed_values in observed_cols.items():
+                        responses = responses.filter(
+                            polars.col(col).is_in(observed_values)
+                        )
+
+                    pivoted = responses.collect().pivot(
+                        on="realization",
+                        index=["response_key", *response_cls.primary_key],
+                        values="values",
+                        aggregate_function="mean",
+                    )
+
+                    if pivoted.is_empty():
+                        joined = observations_for_type.with_columns(
+                            polars.Series(
+                                str(real),
+                                [np.nan] * len(observations_for_type),
+                                dtype=polars.Float32,
+                            )
+                        )
```

Comment:
Why do we need to fill it with nans?

Reply:
Since it is per realization, each round in this loop is constructing a col for response values per real. If there were no responses for realization `N`, we still want the nans in the column for it I suppose?